### PR TITLE
refactor: collapse encrypt preview submission middleware

### LINKED ISSUE #1435: Migrate encrypt-submissions.server.controller.js to TS

### PR DESCRIPTION
**Pull request by mantariksh**
Combines the middleware for the endpoint which handles storage mode preview submissions.

Closes #1435 

## Other improvements
While writing integration tests, I noticed a couple of issues with the Joi validation shared by the actual submission and preview submission endpoints:
1. `helper.error` results in Joi returning an error message, because they require a custom string (e.g. `'any.custom'`) to be passed as the first argument, rather than our own error message.
2. The `encryptedFile` key in individual attachments was not a required key, even though it should be.

These two issues have been fixed in `encrypt-submission.middleware.ts`.

## Tests
- [ ] Create a storage mode form and add an email field with email confirmation. Submit it in preview mode. You should not be able to see the response in the data tab, but you should receive the email confirmation.
- [ ] Change the authType to SP and submit in preview mode.
- [ ] Change the authType to CP and submit in preview mode.

### CODE REVIEW

File: src/app/modules/submission/encrypt-submission/__tests__/encrypt-submission.service.spec.ts [hunk 1]
```diff
@@ -31,7 +35,49 @@ import {
...
+      admin: new ObjectId(),
+      _id: new ObjectId(),
+      title: 'mock title',
+      getUniqueMyInfoAttrs: () => [],
+      authType: 'NIL',
+    } as unknown) as IPopulatedEncryptedForm
+    const MOCK_ENCRYPTED_CONTENT = 'mockEncryptedContent'
+    const MOCK_VERIFIED_CONTENT = 'mockVerifiedContent'
+    const MOCK_VERSION = 1
+    const MOCK_ATTACHMENT_METADATA = new Map<string, string>()
```
Comment (karrui) [inline src/app/modules/submission/encrypt-submission/__tests__/encrypt-submission.service.spec.ts:56]: nit: can make it cleaner with

```ts
const MOCK_ATTACHMENT_METADATA = new Map([['a', 'b']])
```

File: src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts [hunk 2]
```diff
@@ -348,3 +352,30 @@ export const checkFormIsEncryptMode = (
...
+ * @returns Encrypted submission document which has not been saved to database
+ */
+export const createEncryptSubmissionWithoutSave = ({
+  form,
+  encryptedContent,
+  version,
+  attachmentMetadata,
+  verifiedContent,
+}: SaveEncryptSubmissionParams): IEncryptedSubmissionSchema => {
+  return new EncryptSubmissionModel({
```
Comment (karrui) [inline src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts:372]: nit: should we use EncryptSubmissionModel.create instead for type checking?
(or does the new mongoose update now add type safety for this style of instantiation too?)

File: src/app/routes/public-forms.server.routes.js [hunk 3]
```diff
@@ -155,54 +155,13 @@ module.exports = function (app) {
...
-    forms.formById,
-    EncryptSubmissionController.handleEncryptedSubmission,
-  )
+  app
+    .route('/v2/submissions/encrypt/:formId([a-fA-F0-9]{24})')
+    .post(
+      limitRate({ max: rateLimitConfig.submissions }),
+      CaptchaFactory.validateCaptchaParams,
+      EncryptSubmissionMiddleware.validateEncryptSubmissionParams,
+      forms.formById,
```
Comment (karrui) [inline src/app/routes/public-forms.server.routes.js:164]: I assume this is out of scope and that is why its not inlined yet

File: src/types/api/encrypt-submission.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { FieldResponse } from '../response'
+
+export type EncryptSubmissionBody = {
```
Comment (karrui) [inline src/types/api/encrypt-submission.ts:3]: nit: Should we append `Dto`?

File: src/types/api/encrypt-submission.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  responses: FieldResponse[]
+  encryptedContent: string
+  attachments?: EncryptedAttachments
+  isPreview: boolean
+  version: number
+}
+
+export type EncryptedAttachments = {
```
Comment (karrui) [inline src/types/api/encrypt-submission.ts:11]: samesies

**Review by karrui — APPROVED** [review]
lgtm 🎉

File: src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts (hunk 2 above)
Comment (mantariksh) [reply src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts:372]: the point of this function is _not_ to save the submission to the database, since this is a preview submission (hence the `WithoutSave` in the function name hahaha

File: src/app/routes/public-forms.server.routes.js (hunk 3 above)
Comment (mantariksh) [reply src/app/routes/public-forms.server.routes.js:164]: yep correct, we can delete it once #1398 is merged and we migrate this file to TS

File: src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts (hunk 2 above)
Comment (karrui) [reply src/app/modules/submission/encrypt-submission/encrypt-submission.service.ts:372]: ah, missed it

File: src/app/modules/form/admin-form/admin-form.controller.ts [hunk 6]
```diff
@@ -60,7 +55,7 @@ const logger = createLoggerWithLabel(module)
 export const handleListDashboardForms: RequestHandler = async (req, res) => {
   const authedUserId = (req.session as Express.AuthedSession).user._id
 
-  return getDashboardForms(authedUserId)
+  return AdminFormService.getDashboardForms(authedUserId)
```
[the remark this answers is not in the document]
Comment (mantariksh) [reply src/app/modules/form/admin-form/admin-form.controller.ts:58]: out of scope imo, dyou wanna do this in a separate PR?
